pkg/storegateway/bucket.go

```diff
@@ -1288,6 +1274,95 @@ func (s *BucketStore) LabelValues(ctx context.Context, req *storepb.LabelValuesR
 	}, nil
 }
 
+func (s *BucketStore) synchronousSeriesSet(
+	ctx context.Context,
+	req *storepb.SeriesRequest,
+	blocks []*bucketBlock,
+	chunkBytes *pool.BatchBytes,
+	shardSelector *sharding.ShardSelector,
+	matchers []*labels.Matcher,
+	chunksLimiter ChunksLimiter,
+	seriesLimiter SeriesLimiter,
+) (storepb.SeriesSet, *hintspb.SeriesResponseHints, *queryStats, func(), error) {
+	resHints := &hintspb.SeriesResponseHints{}
+	mtx := sync.Mutex{}
+	seriesSets := make([]storepb.SeriesSet, 0, len(blocks))
+	g, ctx := errgroup.WithContext(ctx)
+	stats := &queryStats{}
+	cleanups := make([]func(), 0, len(blocks)*2)
+
+	for _, b := range blocks {
+		b := b
+
+		// Keep track of queried blocks.
+		resHints.AddQueriedBlock(b.meta.ULID)
+
+		var chunkr *bucketChunkReader
+		indexr := b.indexReader()
+		if !req.SkipChunks {
+			chunkr = b.chunkReader(ctx, chunkBytes)
+			cleanups = append(cleanups, func() {
+				// We must keep the readers open until all their data has been sent.
+				runutil.CloseWithLogOnErr(s.logger, chunkr, "series block")
+			})
+		}
+
+		cleanups = append(cleanups, func() {
+			// Defer all closes to the end of Series method.
+			runutil.CloseWithLogOnErr(s.logger, indexr, "series block")
+		})
+
+		// If query sharding is enabled we have to get the block-specific series hash cache
+		// which is used by blockSeries().
+		var blockSeriesHashCache *hashcache.BlockSeriesHashCache
+		if shardSelector != nil {
+			blockSeriesHashCache = s.seriesHashCache.GetBlockCache(b.meta.ULID.String())
+		}
+		g.Go(func() error {
+			var (
+				pstats *safeQueryStats
+				part   storepb.SeriesSet
+				err    error
+			)
+
+			part, pstats, err = blockSeries(
+				ctx, indexr, chunkr, matchers, shardSelector, blockSeriesHashCache, chunksLimiter, seriesLimiter, req.SkipChunks, req.MinTime, req.MaxTime, req.Aggregates, s.logger)
+
+			if err != nil {
+				return errors.Wrapf(err, "fetch series for block %s", b.meta.ULID)
+			}
+
+			mtx.Lock()
+			seriesSets = append(seriesSets, part)
+			if pstats != nil {
+				stats = stats.merge(pstats.export())
+			}
+			mtx.Unlock()
+
+			return nil
+		})
+	}
+
+	s.mtx.RUnlock()
+	cleanup := func() {
+		for _, c := range cleanups {
+			c()
+		}
+	}
```

Comment:
the cleanups need to happen both when we return an error and when we don't. Since the returns are two, _this_ (lines 1347-1351) is one of the latest common points for the two code paths

Reply:
Add a comment to explain it please.